File: vol1/ch3/LeetCode32.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        n=len(s)
        if n==0:return 0
        ans=0
        st=[]
        st.append(-1)
        for i in range(0,n):
            if s[i]=='(':
                st.append(i)
            else:
                st.pop()
                if len(st)==0:
                    st.append(i)
                else:
                    ans=max(ans,i-st[-1])
        return ans

File: vol1/ch3/LeetCode32.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int n=s.size();
        if(n==0) return 0;
        int ans=0;
        stack<int> st;
        st.push(-1);
        for (int i=0;i<n;i++) {
            if (s[i]=='(') {
                st.push(i);
            }
            else {
                st.pop();
                if (st.empty())
                    st.push(i);							//更新栈底为最后一个没有被匹配的右括号的下标
                else										//找到一个有效括号子串
                    ans=max(ans,i-st.top());		//其长度=i-st.top()
            }
        }
        return ans;
    }
};
